include/navigation/insilico_rotater.h

```diff
@@ -51,36 +50,36 @@ struct InSilicoRotater
             BOB_ASSERT(beginRoll < endRoll);
             BOB_ASSERT((distance(endRoll, beginRoll) % scanStep) == 0);
 
-            const auto index = toIndex(beginRoll);
-            rollImage(image, m_Image, index);
+            for (auto &image : m_ScratchImages.elements()) {
+                image.create(unwrapRes, CV_8UC1);
+            }
             if (!maskImage.empty()) {
-                rollImage(maskImage, m_MaskImage, index);
+                for (auto &mask : m_ScratchMaskImages.elements()) {
+                    mask.create(unwrapRes, maskImage.type());
+                }
             }
-        }
+       }
 
         template<class Func>
-        void rotate(Func func)
+        void rotate(Func func) const
         {
-            auto i = m_BeginRoll;
-            while (true) {
-                func(m_Image, m_MaskImage, distance(m_BeginRoll, i));
-
-                i += m_ScanStep;
-                if (i >= m_EndRoll) {
-                    break;
-                }
-
+            #pragma omp parallel for
```

Comment:
I don't see how this loop can be parallelised - you're rolling the same image pixel by pixel so each iteration depends on the previous one.

Reply:
We might have done that at some point in the distant past, but we were using a scratch variable to store the rotated image here even before I started mangling things :smile: .

I actually did what you suggested initially (i.e. parallelising the inner loop which iterates over the different snapshots), but it doesn't perform as well, I guess a) because the rotations are still being done on a single thread and b) because whatever OpenMP synchronisation magic happens when the loop finishes would have to happen ``numRotations`` times. The other advantage to parallelising the outer loop is that any other algorithms using ``InSilicoRotater`` also get a speedup (Infomax now also runs much faster, but I didn't print the figures).

Reply:
Oh yeah, you're totally right - seems I've forgotten how this code works. Each iteration of this loop i.e. thread makes a rotated version of the image and tests it against all the snapshots. This is entirely where it should be parallelised.